README.md
# hetsar: R package to estimate spatial autoregressive models with heterogeneous coefficient

Estimation and inference of heterogeneous spatial autoregressive (HSAR) panel
data models. These are models where the spatial lag coefficients are allowed to
differ over the cross-section units, in addition to the fixed effects generally
allowed for in the literature. The model also features weakly exogenous
regressors, such as lagged values of the dependent variable and heteroskedastic
error. The estimation is performed via quasi maximum-likelihood. See Aquaro,
Bailey and Pesaran (2021) for technical details.

## Installation

```r
# install release version from CRAN
install.packages("hetsar")

# install development version from GitHub
devtools::install_github("maquaro/hetsar")
```

## Usage

```r
library("hetsar")

data(hetsarDataDemo)
df_data <- hetsarDataDemo[["data"]]
m_C <- hetsarDataDemo[["network_matrix"]]

# row normalise
m_W <- m_C / rowSums(m_C)

# y = Wy + 1 + x + Wx + err
# other model specifications available, see `?hetsar`
out <- hetsar(
  data = df_data, 
  W = m_W,
  indices = c("id", "time"),
  dependent = "y",
  explanatory = "x",
  space_lags = "x")

summary(out)
summary(out, MG = TRUE)
```

## Vignettes

Coming soon.

## Other software

**hetsar** is also available in
MATLAB (coming soon),
[Python](https://pypi.org/project/hetsar/), and
[Stata](https://ideas.repec.org/c/boc/bocode/s458926.html).

## References

M. Aquaro, N. Bailey and M. H. Pesaran (2021).
"Estimation and inference for spatial models with heterogeneous coefficients: An application to US house prices".
Journal of Applied Econometrics 36(1): 18-44.
doi:https://doi.org/10.1002/jae.2792


NEWS
NEWS for hetsar package

VERSION 0.0.1 (2021-09-xx)

-   Initial upload to CRAN.

VERSION 0.0.0.9000 (2021-09-21)

-   Initial upload to GitHub.


R/fn_gr.R
#-- Gradient of the function: phi(vtheta):=-(logL(vtheta) / T) --#
fn_gr <- function(v_theta, l_data) {
     # retrieve data
     m_y  <- l_data[["m_y" ]]
     m_ys <- l_data[["m_ys"]]
     l_x  <- l_data[["l_x" ]]
     m_W  <- l_data[["m_W" ]]

     # sample size
     N <- nrow(m_y)
     TT <- ncol(m_y)
     K <- length(l_x)

     v_psi   <- v_theta[1:N] # (N,)
     v_beta  <- v_theta[(N + 1):(N + (K * N))]; # (KN,) x 1
     v_sgmsq <- v_theta[(N + (K * N) + 1):(N + (N * K) + N)] # (N,)

     # reshape beta
     m_beta_tr <- matrix(v_beta, K, N) # (K,N)
     m_beta <- t(m_beta_tr) # (N,K)

     v_sgm4h <- v_sgmsq^2
     v_sgm6h <- v_sgmsq^3

     # compute residuals
     m_beta_times_x <- matrix(0, N, TT) # 0s needed for recursive sum; (N,T) with generic element \vbeta_{i}'\vx_{it}
     for (k in 1:K) {
          m_x_k <- l_x[[k]] # (N,T)
          v_beta_k <- m_beta[, k] # (N,)
          m_beta_times_x_k <- m_x_k * v_beta_k #!! (N,T) "times" (N,1) = (N,T)
          m_beta_times_x <- m_beta_times_x + m_beta_times_x_k
     }
     m_psi_times_ys <- m_ys * v_psi #!! (N,T) "times" (N,1) = (N,T)
     m_eps <- m_y - m_psi_times_ys - m_beta_times_x # N x T
     v_ssr <- rowSums(m_eps^2) # (N,) sum of squared residuals: sum_t (eps_it)^2 !!crossprod()??

     # standardize
     sssr <- sum(v_ssr / v_sgmsq) # sum of squared standardized-residuals: sum_i sum_t (eps_it / sgm_i)^2

     m_Psi <- diag(v_psi, nrow = N) # (N,N)
     m_A <- diag(N) - (m_Psi %*% m_W)

     det_mA <- det(m_A)
     if (det_mA <= 0) {
          stop("`I - Psi W` is not invertible. Check that the parameter space is defined correctly.")
     }

     # first derivative
     ## v_dphi_dvpsi
     m_Q <- m_W %*% solve(m_A) #!! IS THERE A BETTER WAY TO WRITE THIS LINE?
     v_dphi_dvpsi <- diag(m_Q) - (rowSums(m_ys * m_eps) / v_sgmsq / TT) #!! stopifnot(dim(m_Q) > 1)

     ## v_dphi_dvbeta
     m_X_times_eps_divided_sgmsq <- matrix(NA_real_, N, K)
     m_eps_divided_sgmsq <- m_eps / v_sgmsq # (N,T)
     for (k in 1:K) {
          m_x_k <- l_x[[k]] # (N,T)
          m_x_k_times_eps_divided_sgmsq <- m_x_k * m_eps_divided_sgmsq # (N,T)
          m_X_times_eps_divided_sgmsq[, k] <- rowSums(m_x_k_times_eps_divided_sgmsq) # (N,)
     }
     m_X_times_eps_divided_sgmsq_tr <- t(m_X_times_eps_divided_sgmsq) # (K,N)
     v_dphi_dvbeta <- -c(m_X_times_eps_divided_sgmsq_tr) / TT # (NK,)

     ## v_dphi_dvsgmsq
     v_dphi_dvsgmsq <- (1 / 2 / v_sgmsq) - (v_ssr / v_sgm4h / TT / 2)
     v_gradient <- c(v_dphi_dvpsi,
                     v_dphi_dvbeta,
                     v_dphi_dvsgmsq)

     # return
     v_gradient
}


R/fn_hsar.R
#-- wrapper: write the estimator as a function of the dataset, only --#
fn_hsar <- function(l_data, l_bounds) {
     # -------------------------------------------------------------------------- #
     # l_data is a list containing:
     # m_y: (N,T) matrix
     # l_x: list containing K matrices of covariates of order (N,T) (it may or may note include the intercept)
     # m_W: (N,N) matrix

     # -------------------------------------------------------------------------- #
     N  <- nrow(l_data[["m_y"]])
     TT <- ncol(l_data[["m_y"]])
     K <- length(l_data[["l_x"]])

     # add y*:=Wy
     l_data[["m_ys"]] <- l_data[["m_W"]] %*% l_data[["m_y"]] # (N,T)
     

     # -------------------------------------------------------------------------- #
     # constructs v_theta_ini
     l_theta_ini <- NULL #cc#
     v_theta_ini <- vector("numeric", N * (K + 2))
     if (is.null(l_theta_ini)) {
          v_theta_ini[1:(N + (K * N))] <- 0
          v_theta_ini[(N + (K * N) + 1):(N + (N * K) + N)] <- 1
     } else {
          v_psi_ini   <- l_theta_ini[["psi"  ]]
          m_beta_ini  <- l_theta_ini[["beta" ]]
          v_sgmsq_ini <- l_theta_ini[["sqmsq"]]

          ## reshape m_beta_ini
          m_beta_ini_tr <- t(m_beta_ini) # (K,N)
          v_beta_ini <- c(m_beta_ini_tr) # (NK,)

          ## fill v_theta_ini
          v_theta_ini[1:N]                                 <- v_psi_ini   # (N,)
          v_theta_ini[(N + 1):(N + (K * N))]               <- v_beta_ini  # (KN,)
          v_theta_ini[(N + (K * N) + 1):(N + (N * K) + N)] <- v_sgmsq_ini # (N,)
     }

     # -------------------------------------------------------------------------- #
     l_out <- stats::optim(
          par     = v_theta_ini, 
          fn      = fn_ofun,
          gr      = fn_gr,
          l_data  = l_data,
          method  = "L-BFGS-B",
          lower   = l_bounds[[1]], # (N(K+2),)
          upper   = l_bounds[[2]], # (N(K+2),)
          control = list(maxit = 1e6))
     v_theta  <- l_out[["par"]]
     #print(l_out)

     v_psi   <- v_theta[1:N] # (N,)
     v_beta  <- v_theta[(N + 1):(N + (K * N))]; # (KN,)
     v_sgmsq <- v_theta[(N + (K * N) + 1):(N + (N * K) + N)] # (N,)
     #print(range(v_sgmsq))

     # reshape beta
     m_beta_tr <- matrix(v_beta, K, N) # (K,N)
     m_beta <- t(m_beta_tr) # (N,K)

     # pack all estimates into a matrix with names
     m_theta <- matrix(NA_real_, N, K + 2)
     m_theta[, 1]         <- v_psi
     m_theta[, 2:(K + 1)] <- m_beta
     m_theta[, (K + 2)]   <- v_sgmsq
     #browser()
     #print(m_theta)

     # -------------------------------------------------------------------------- #
     # residuals

     m_y  <- l_data[["m_y" ]]
     m_ys <- l_data[["m_ys"]]
     l_x  <- l_data[["l_x" ]]
     m_W  <- l_data[["m_W" ]]
     N <- nrow(m_y)
     TT <- ncol(m_y)
     K <- length(l_x)

     m_beta_times_x <- matrix(0, N, TT) # 0 needed for recursive sum; (N x T) with generic element \vbeta_{i}'\vx_{it}
     for (k in 1:K) {
          m_x_k <- l_x[[k]] # (N,T)
          v_beta_k <- m_beta[, k] # (N,)
          m_beta_times_x_k <- m_x_k * v_beta_k #!!
          m_beta_times_x <- m_beta_times_x + m_beta_times_x_k
     }
     m_psi_times_ys <- m_ys * v_psi #!! N x T
     m_yhat_naive <- m_psi_times_ys + m_beta_times_x
     m_resid_naive <- m_y - m_yhat_naive # (N,T)

     m_Psi <- diag(v_psi, nrow = N) # (N,N)
     m_A <- diag(N) - (m_Psi %*% m_W)
     m_yhat_rform <- solve(m_A, m_beta_times_x)
     m_resid_rform <- m_y - m_yhat_rform # (N,T)

     # -------------------------------------------------------------------------- #
     # variance
     l_variance <- fn_var_ml(m_theta, l_data) # (2,)

     # -------------------------------------------------------------------------- #
     # return
     l_return <- list()
     l_return[["theta"]] <- m_theta # (3,)

     l_return[["standard"]] <- l_variance[["standard"]]
     l_return[["sandwich"]] <- l_variance[["sandwich"]]

     l_return[["resid_rform"]] <- m_resid_rform
     l_return[["resid_naive"]] <- m_resid_naive

     l_return[["fval"]] <- l_out[["value"]]
     l_return[["exitflag"]] <- l_out[["convergence"]]
     l_return[["optim.msg"]] <- l_out[["message"]]

     l_return
}


R/fn_ofun.R
#-- Objective function: phi(vtheta):=-(logL(vtheta) / T) --#
fn_ofun <- function(v_theta, l_data) {
     # retrieve data
     m_y  <- l_data[["m_y" ]]
     m_ys <- l_data[["m_ys"]]
     l_x  <- l_data[["l_x" ]]
     m_W  <- l_data[["m_W" ]]

     # sample size
     N <- nrow(m_y)
     TT <- ncol(m_y)
     K <- length(l_x)

     v_psi   <- v_theta[1:N] # (N,)
     v_beta  <- v_theta[(N + 1):(N + (K * N))]; # (KN,) x 1
     v_sgmsq <- v_theta[(N + (K * N) + 1):(N + (N * K) + N)] # (N,)

     # reshape beta
     m_beta_tr <- matrix(v_beta, K, N) # (K,N)
     m_beta <- t(m_beta_tr) # (N,K)

     v_sgm4h <- v_sgmsq^2
     v_sgm6h <- v_sgmsq^3

     # compute residuals
     m_beta_times_x <- matrix(0, N, TT) # 0 needed for recursive sum; (N x T) with generic element \vbeta_{i}'\vx_{it}
     for (k in 1:K) {
          m_x_k <- l_x[[k]] # (N,T)
          v_beta_k <- m_beta[, k] # (N,)
          m_beta_times_x_k <- m_x_k * v_beta_k #!!
          #browser()
          #print(dim(m_beta_times_x))
          #print(dim(m_beta_times_x_k))
          m_beta_times_x <- m_beta_times_x + m_beta_times_x_k
     }
     m_psi_times_ys <- m_ys * v_psi #!! N x T
     m_eps <- m_y - m_psi_times_ys - m_beta_times_x # N x T
     v_ssr <- rowSums(m_eps^2) # (N,) sum of squared residuals: sum_t (eps_it)^2 !!crossprod()??

     # standardize
     sssr <- sum(v_ssr / v_sgmsq) # sum of squared standardized-residuals: sum_i sum_t (eps_it / sgm_i)^2

     m_Psi <- diag(v_psi, nrow = N) # (N,N)
     m_A <- diag(N) - (m_Psi %*% m_W)

     det_mA <- det(m_A)
     if (det_mA <= 0) {
          stop("`I - Psi W` is not invertible. Check that the parameter space is defined correctly.")
     }
     constant <- log(2 * pi) * N / 2
     first_part <- -log(det_mA)
     secon_part <- sum(log(v_sgmsq)) / 2
     third_part <- sssr / TT / 2
     ofun <- constant + first_part + secon_part + third_part
     #print(ofun)
     #print(first_part)
     #print(secon_part)
     #print(third_part)
     #print(v_sgmsq)
     #print(sssr)
     #print(v_ssr)
     #print(v_sgmsq)

     # return
     ofun
}


R/hetsar.R
#' Heterogeneous SAR estimation
#'
#' Estimation of heterogeneous spatial autoregressive models
#'
#' @param data A `data.frame`, in tidy format. The panel _must_ be balanced.
#' @param W Spatial weight `matrix`. Rows and columns should be ordered such that colnames(W)==rowanems(W)==unique(df$id).
#' @param indices Character vector of length 2 with the names of the columns in `data` containing the unit-identifier (`indices[[1]]`) and the time-identifier (`indices[[2]]`).
#' @param dependent Name of the dependent variable.
#' @param explanatory Character vector with the names of the explanatory variables, excluding: (a) the spatial lag of the dependent variable `Wy`, included by default; (b) the intercept; (c) space/time-lags. See Details.
#' @param nocons Logical: Shall the constant term (i.e. the intercept) be excluded?
#' @param time_lags Character vector with the names of the time-lagged variables.
#' @param space_lags Character vector with the names of the space-lagged variables.
#' @param time_space_lags Character vector with the names of the time- and space-lagged variables.
#' @param verbose Character vector. Verbosely print some extra information relative to the model specification (`verbose="model"`), to the bounds (`verbose="bounds"`), to the optimisation procedure (`verbose="optimisation"`), or to all the above (`verbose="all"). Default is silent (`verbose="silent"`).
#' @param bounds (Advanced) Named-`list` specifying the parameter space, see "Bounds".
#'
#' @return
#' TODO An object (`list`) of class `hetsar` with the following elements:
#' \item{coefficients}{A matrix of estimated coefficients. !! COLNAMES}
#' \item{Std.err.sandwich}{A matrix with standard errors computed using the sandwich formula.}
#' \item{Std.err.standard}{A matrix with standard errors computed using the standard formula.}
#' \item{N}{The number of spatial units.}
#' \item{T}{The number of time periods.}
#' \item{NT}{The number of observations.}
#' 
#' @section Model specification:
#' TODO
#' 
#' @section Bounds:
#' TODO
#'
#' @seealso
#' TODO See also \code{\link[hetsar]{summary.hetsar}} to see the results.
#'
#' @author
#' Michele Aquaro
#'
#' @references
#' M. Aquaro, N. Bailey and M. H. Pesaran (2021). "Estimation and inference for spatial models with heterogeneous coefficients: An application to US house prices". Journal of Applied Econometrics 36(1): 18-44. doi:https://doi.org/10.1002/jae.2792
#'
#' @examples
#'
#' data(hetsarDataDemo)
#' df_data <- hetsarDataDemo[["data"]]
#' m_C <- hetsarDataDemo[["network_matrix"]]
#' m_W <- m_C / rowSums(m_C) # row normalise
#'
#' out <- hetsar(
#'   data = df_data, 
#'   W = m_W,
#'   indices = c("id", "time"),
#'   dependent = "y",
#'   explanatory = "x",
#'   space_lags = "x")
#'
#' summary(out)
#' summary(out, MG = TRUE)
#'
#' @export
hetsar <- function(
     data, 
     W, 
     indices, # can [[1]] be a character? Can [[2]] be a date?
     dependent,
     explanatory = NULL,
     nocons = FALSE,
     time_lags = NULL,
     space_lags = NULL,
     time_space_lags = NULL,
     verbose = "silent",
     bounds = NULL) {

     # -------------------------------------------------------------------------- #
     # rename
     df_data <- data
     m_W <- W
     #!! rm(list(data, m_W)) ?

     # ========================================================================== #
     # CHECKS

     # -------------------------------------------------------------------------- #
     # checks type/class

     if (!any(class(df_data) == "data.frame")) {
          stop("`data` must be an object of class 'data.frame'")
     }

     if (!any(class(m_W) == "matrix")) {
          stop("`W` must be an object of class 'matrix'")
     }

     if (!typeof(dependent) == "character") {
          stop("`dependent` must be an object of type 'character'")
     }

     if (!is.null(explanatory)) {
          if (!typeof(explanatory) == "character") {
               stop("`explanatory` must be an object of type 'character'")
          }
     }

     if (!typeof(nocons) == "logical") {
          stop("`nocons` must be an object of type 'logical'")
     }

     if (!is.null(time_lags)) {
          if (!typeof(time_lags) == "character") {
               stop("`time_lags` must be an object of type 'character'")
          }
     }

     if (!is.null(space_lags)) {
          if (!typeof(space_lags) == "character") {
               stop("`space_lags` must be an object of type 'character'")
          }
     }

     if (!is.null(time_space_lags)) {
          if (!typeof(time_space_lags) == "character") {
               stop("`time_space_lags` must be an object of type 'character'")
          }
     }

     if (!typeof(verbose) == "character") {
          stop("`verbose` must be an object of type 'character'")
     }

     if (!is.null(bounds)) {
          if (!typeof(bounds) == "list") {
               stop("`bounds` must be an object of type 'list'")
          }
     }

     # -------------------------------------------------------------------------- #
     # other checks

     if (length(dependent) > 1) {
          stop("`dependent` _must_ be a character vector of length 1 (i.e. only one outcome variable is allowed)")
     }

     if (length(indices) != 2) {
          stop("One of the two indices is missing (`indices` _must_ be a character vector of length 2)")
     }

     # -------------------------------------------------------------------------- #
     # checks on `df_data`

     # 1. are all variables to be used in data?
     v_variables <- c(indices, dependent, explanatory, time_lags, space_lags, time_space_lags)
     v_bool <- v_variables %in% names(df_data)
     if (!all(v_bool)) {
          # location of the unmatched column name(s)
          v_ind <- which(!v_bool)
          stop(sprintf("%s not found in `data`", paste(v_variables[v_ind], collapse = ", ")))
     }
     df_data <- df_data[, v_variables] # copy

     # 2. remove NAs
     NT_old <- nrow(df_data)
     df_data <- stats::na.omit(df_data)
     NT_new <- nrow(df_data)
     if (NT_new < NT_old) {
          cat(sprintf("  Removed %d row(s) with NAs\n", NT_old - NT_new))
     }

     # 3. balanced?
     v_NT_idx_i <- df_data[[indices[[1L]]]] # (NT,)
     v_NT_idx_t <- df_data[[indices[[2L]]]]
     N. <- length(unique(v_NT_idx_i))
     T. <- length(unique(v_NT_idx_t))
     NT <- N. * T.
     if (nrow(df_data) != NT) {
          stop("`data` is not balanced (i.e. nrow(df_data)!=NT)")
     }

     # 4. sort
     df_data <- df_data[order(v_NT_idx_t, v_NT_idx_i), ]

     # 5. no "L." or "W." or "L.W." in the names
     v_colnames <- names(df_data)
     v_ind <- grep("L.", substr(v_colnames, 1, 2), fixed = TRUE)
     if (length(v_ind) > 0) {
          stop(sprintf("Rename column(s) %s in `data` (pls use something different than `L.foo`, see Documentation)", paste(v_colnames[v_ind], collapse = ", ")))
     }
     v_ind <- grep("W.", substr(v_colnames, 1, 2), fixed = TRUE)
     if (length(v_ind) > 0) {
          stop(sprintf("Rename column(s) %s in `data` (pls use something different than `W.foo`, see Documentation)", paste(v_colnames[v_ind], collapse = ", ")))
     }

     # -------------------------------------------------------------------------- #
     # checks on `m_W`
     
     if (nrow(m_W) != ncol(m_W)) {
          stop("`W` must be a square matrix")
     }

     if (any(c(m_W) < 0)) {
          stop("negative weights in `m_W` are not allowed")
     }

     if (!all(rowSums(m_W) > 0)) {
          stop("each spatial unit in `m_W` must have at least 1 neighbour")
     }

     if (nrow(m_W) != N.) {
          stop("`df_data` and `m_W` must have the same number of spatial units N")
     }

     # -------------------------------------------------------------------------- #
     # checks `l_x` _must_ have at least one element
     v_bool <- c(
          nocons,
          is.null(explanatory),
          is.null(time_lags),
          is.null(space_lags),
          is.null(time_space_lags))
     if (all(v_bool)) {
          stop("'y = Wy + e' not coded. Please add at least one variable on the rhs of the equation")
     }

     # -------------------------------------------------------------------------- #
     # checks on bounds

     # construct the full list of names that can appear in bounds
     v_name_bounds_universe <- c(paste0("W.", dependent), "sigmasq")
     if (!nocons) {
          v_name_bounds_universe <- c(v_name_bounds_universe, "cons")
     }
     if (!is.null(explanatory)) {
          v_name_bounds_universe <- c(v_name_bounds_universe, explanatory)
     }
     if (!is.null(time_lags)) {
          v_name_bounds_universe <- c(v_name_bounds_universe, paste0("L.", time_lags))
     }
     if (!is.null(space_lags)) {
          v_name_bounds_universe <- c(v_name_bounds_universe, paste0("W.", space_lags))
     }
     if (!is.null(time_space_lags)) {
          v_name_bounds_universe <- c(v_name_bounds_universe, paste0("L.W.", time_space_lags))
     }
     # (do I need to worry about duplicates? Perhaps not)
     v_name_bounds_universe <- unique(v_name_bounds_universe)
     if (!is.null(bounds)) {
          v_bool <- names(bounds) %in% v_name_bounds_universe
          if (!all(v_bool)) {
               stop(sprintf("%s %s",
                    "Some names in `bounds` seem to be mispelled (i.e. `names(bounds)`).\n  Allowed names in `bounds` are: ",
                    paste(v_name_bounds_universe, collapse = ", ")))
          }
     }

     # ========================================================================== #
     # LONG TO WIDE
     
     # -------------------------------------------------------------------------- #
     # explanatory variables
     l_x <- list()

     ## constant
     if (!nocons) {
          ## remove t=1 if there are time-lags
          if (length(time_lags) > 0 || length(time_space_lags) > 0) {
               l_x[["cons"]] <- matrix(1, N., T. - 1)
          } else {
               l_x[["cons"]] <- matrix(1, N., T.)
          }
     }

     ## non-lagged explanatory variables
     if (!is.null(explanatory)) {
          for (x in explanatory) {
               m_x <- matrix(df_data[[x]], N., T.)

               ## remove t=1 if there are time-lags
               if (length(time_lags) > 0 || length(time_space_lags) > 0) {
                    l_x[[x]] <- m_x[, 2:T.]
               } else {
                    l_x[[x]] <- m_x
               }
          }
     }

     ## time-lagged explanatory variables
     if (!is.null(time_lags)) {
          for (x in time_lags) {
               m_x <- matrix(df_data[[x]], N., T.)
               l_x[[paste0("L.", x)]] <- m_x[, 1:(T. - 1)]
          }
     }

     ## space-lagged explanatory variables
     if (!is.null(space_lags)) {
          for (x in space_lags) {
               m_x <- matrix(df_data[[x]], N., T.)
               m_Wx <- m_W %*% m_x

               ## remove t=1 if there are time-lags
               if (length(time_lags) > 0 || length(time_space_lags) > 0) {
                    l_x[[paste0("W.", x)]] <- m_Wx[, 2:T.]
               } else {
                    l_x[[paste0("W.", x)]] <- m_Wx
               }
          }
     }

     ## time-space-lagged explanatory variables
     if (!is.null(time_space_lags)) {
          for (x in time_space_lags) {
               m_x <- matrix(df_data[[x]], N., T.)
               m_Wx <- m_W %*% m_x
               l_x[[paste0("L.W.", x)]] <- m_Wx[, 1:(T. - 1)]
          }
     }

     # -------------------------------------------------------------------------- #
     # dependent variable
     m_y <- matrix(df_data[[dependent]], N., T.) 

     # remove t=1 if there are time-lags
     if (length(time_lags) > 0 || length(time_space_lags) > 0) {
          m_y <- m_y[, 2:T.]
     }

     # -------------------------------------------------------------------------- #
     # pack all data into a list
     l_data <- list()
     l_data[["m_y"]] <- m_y
     l_data[["l_x"]] <- l_x
     l_data[["m_W"]] <- m_W
     #browser()

     # ========================================================================== #
     # bounds

     K <- length(l_x)
     N <- N.

     # character vector with parameter names
     v_name_parameter <- c(
          paste0("W.", dependent),
          names(l_x),
          "sigmasq")

     # if `bounds` is NULL assign it to a named-list
     if (is.null(bounds)) {
          bounds <- vector("list", length(v_name_parameter)) # (K+2,)
          names(bounds) <- v_name_parameter
     }

     bnd <- 0.995 #cc#

     # default lower/upper-bounds for W.y
     if (is.null(bounds[[paste0("W.", dependent)]])) {
          bounds[[paste0("W.", dependent)]] <- c(-bnd, bnd)
     }

     # default lower/upper-bounds for sigmasq
     if (is.null(bounds[["sigmasq"]])) {
          bounds[["sigmasq"]] <- c(0.01, Inf)
     }

     # default lower/upper-bounds for L.y
     if ( (dependent %in% time_lags) &&  is.null(bounds[[paste0("L.", dependent)]]) ) {
          bounds[[paste0("L.", dependent)]] <- c(-bnd, bnd)
     }

     # default lower/upper-bounds for L.W.y
     if ( (dependent %in% time_lags) &&  is.null(bounds[[paste0("L.W.", dependent)]]) ) {
          bounds[[paste0("L.W.", dependent)]] <- c(-bnd, bnd)
     }

     # do the rest of the variables (e.g. `x`, `W.x`)
     for (np in v_name_parameter) { # mnemonic: Name Parameter
          if (is.null(bounds[[np]])) {
               # default lower/upper-bounds for x
               bounds[[np]] <- c(-Inf, Inf)
          } else {
               next
          }
     }
     #browser()

     ## from "list" to lower/upper-bound "matrix"
     m_lb_theta <- matrix(NA_real_, N, K + 2)
     m_ub_theta <- matrix(NA_real_, N, K + 2)
     colnames(m_lb_theta) <- v_name_parameter
     colnames(m_ub_theta) <- v_name_parameter
     for (np in v_name_parameter) { # mnemonic: Name Parameter
          vec <- bounds[[np]] # (2,)
          m_lb_theta[, np] <- rep(vec[[1]], N)
          m_ub_theta[, np] <- rep(vec[[2]], N)
     }
     #print(m_lb_theta)
     #print(m_ub_theta)

     ## reshape m_lb_beta
     m_lb_beta <- m_lb_theta[, names(l_x)] # (N,K)
     m_ub_beta <- m_ub_theta[, names(l_x)] # (N,K)
     m_lb_beta_tr <- t(m_lb_beta) # (K,N)
     m_ub_beta_tr <- t(m_ub_beta) # (K,N)
     v_lb_beta <- c(m_lb_beta_tr) # (NK,)
     v_ub_beta <- c(m_ub_beta_tr) # (NK,)

     ## fill v_lb and v_ub
     v_lb <- vector("numeric", N * (K + 2))
     v_ub <- vector("numeric", N * (K + 2))

     v_lb[1:N] <- m_lb_theta[, paste0("W.", dependent)] # (N,)
     v_ub[1:N] <- m_ub_theta[, paste0("W.", dependent)] # (N,)

     v_lb[(N + 1):(N + (K * N))] <- v_lb_beta  # (KN,)
     v_ub[(N + 1):(N + (K * N))] <- v_ub_beta  # (KN,)

     v_lb[(N + (K * N) + 1):(N + (N * K) + N)] <- m_lb_theta[, "sigmasq"] # (N,)
     v_ub[(N + (K * N) + 1):(N + (N * K) + N)] <- m_ub_theta[, "sigmasq"] # (N,)

     l_bounds <- list(v_lb, v_ub)
     #browser()

     # ========================================================================== #
     # VERBOSE

     cat("SAR with heterogeneous coefficients\n\n")

     # formula
     if (verbose == "all" || verbose == "model") {

          # formula
          Kp2 <- length(v_name_parameter)
          cat("Formula (short-hand notation):\n")
          cat(sprintf("  %s = %s + \u03b5\n\n", dependent, paste(v_name_parameter[1:(Kp2 - 1)], collapse = " + ")))
     }

     # bounds
     if (verbose == "all" || verbose == "bounds") {
          cat("Bounds (e.g. parameter space):\n")
          df_temp <- data.frame(variable = v_name_parameter)
          for (np in v_name_parameter) {
               vec <- bounds[[np]] # (2,)
               df_temp[df_temp[["variable"]] == np, "lower_bounds"] <- vec[[1]]
               df_temp[df_temp[["variable"]] == np, "upper_bounds"] <- vec[[2]]
          }
          print(df_temp)
     }

     # ========================================================================== #
     # HSAR
     l_out <- fn_hsar(l_data = l_data, l_bounds = l_bounds)
     m_theta <- l_out[["theta"]]
     m_se_sandwich <- sqrt(l_out[["sandwich"]])
     m_se_standard <- sqrt(l_out[["standard"]])

     # residuals
     m_resid_rform <- l_out[["resid_rform"]]
     m_resid_naive <- l_out[["resid_naive"]]
     if (length(time_lags) > 0 || length(time_space_lags) > 0) {
          # remove t=1 if there are time-lags
          df_resid <- df_data[(N + 1):nrow(df_data), indices] # (NT,2)
     } else {
          df_resid <- df_data[, indices] # (NT,2)
     }
     df_resid[["reduced_form"]] <- c(m_resid_rform)
     df_resid[["naive"]]        <- c(m_resid_naive)

     # report verbosely information on optimisation
     if (verbose == "all" || verbose == "optimisation") {
          cat("\n  Optimisation (see `?optim`):\n")
          cat(sprintf("%s: %d\n", "  - exit flag", l_out[["exitflag"]]))
          cat(sprintf("%s: %f\n", "  - value of the objective function (-1/T)", l_out[["fval"]]))
          cat(sprintf("%s: %s\n", "  - any other message from the solver", l_out[["optim.msg"]]))
     }

     # ========================================================================== #
     # prepare output

     #!! TODO shall I leave these as matrices or shall I transform them into data-frames? 
     #!! What would be the advantage of using data-frames?
     #!! - if I have to re-shuffle columns

     v_id <- unique(df_data[[indices[[1L]]]]) # (N,)
     rownames(m_theta)       <- v_id
     rownames(m_se_sandwich) <- v_id
     rownames(m_se_standard) <- v_id

     colnames(m_theta)       <- v_name_parameter
     colnames(m_se_sandwich) <- v_name_parameter
     colnames(m_se_standard) <- v_name_parameter

     l_return <- list()
     l_return[["coefficients"]] <- m_theta
     l_return[["Std.err.sandwich"]] <- m_se_sandwich
     l_return[["Std.err.standard"]] <- m_se_standard
     l_return[["resid"]] <- df_resid
     l_return[["T"]] <- ncol(m_y)
     l_return[["N"]] <- nrow(m_y)
     l_return[["NT"]] <- nrow(m_y) * ncol(m_y)
     l_return[["fval"]] <- l_out[["fval"]]
     l_return[["exitflag"]] <- l_out[["exitflag"]]
     l_return[["optim.msg"]] <- l_out[["optim.msg"]]

     # add class to the S3 object
     class(l_return) <- "hetsar"

     l_return
}
